src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,146 @@
+ */
+
+import { Emitter, Event } from "../../application/common";
+import { inject, injectable } from "inversify";
+import URI from "../../application/common/uri";
+import { FileSystem, FileChangesEvent } from "../../filesystem/common/filesystem";
+import { FileSystemWatcher } from "../../filesystem/common/filesystem-watcher"
+import * as coreutils from "@phosphor/coreutils";
+
+
+export interface PreferenceChangedEvent {
+    readonly preferenceName: string;
+    readonly newValue?: any;
+    readonly oldValue?: any;
+}
+
+export const PreferenceService = Symbol("PreferenceService")
+export const PreferencePath = Symbol("PreferencePath")
+
+
+/**
+ * Minimal preference API with basic functionalities
+ */
+export interface PreferenceService {
+    // A minimal part of the interface.
+    has(preferenceName: string): Promise<boolean>;
+    get<T>(preferenceName: string): Promise<T | undefined>;
+    readonly onPreferenceChanged: Event<PreferenceChangedEvent>;
+}
+
+export interface PreferenceServiceClient {
+    onDidChangePreference(event: PreferenceChangedEvent): void
+}
+
+@injectable()
+export class JsonPreferenceService implements PreferenceService {
+
+    protected prefs: { [key: string]: any } | undefined; // Preferences cache
+    protected readonly prefsChangedListeners: Emitter<PreferenceChangedEvent> = new Emitter();
+    protected readonly resolveUri: Promise<URI>;
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        // readonly preferencePath: string = ".theia/prefs.json",
+        @inject(PreferencePath) protected readonly preferencePath: string) {
+
+        this.resolveUri = fileSystem.getWorkspaceRoot().then(root =>
+            new URI(root.uri).appendPath(preferencePath)
+        );
+
+        watcher.onFileChanges(event => {
+            this.arePreferencesAffected(event).then((arePrefsAffected) => {
+                if (arePrefsAffected) {
+                    this.reconcilePreferences();
+                }
+            })
+        });
+
+        this.reconcilePreferences();
+    }
+
+    /**
+     * Checks to see if the preference file was modified
+     */
+    protected arePreferencesAffected(event: FileChangesEvent): Promise<boolean> {
+        return this.resolveUri.then(uri =>
+            event.changes.some(c =>
+                c.uri === uri.toString()
+            )
+        );
+    }
+
+    /**
+     * Read preferences
+     */
+    protected reconcilePreferences(): void {
+
+        this.resolveUri.then((uri) => {
+            this.fileSystem.resolveContent(uri.toString()).then(({ stat, content }) => {
+                let newPrefs = JSON.parse(content) // Might need a custom parser because comments and whatnot?
+                this.notifyDifferences(newPrefs);
+            }).catch((error) => {
+                // File does not exist?
```

Comment:
don't do anything if there is no better strategy, messages will be logged anyway